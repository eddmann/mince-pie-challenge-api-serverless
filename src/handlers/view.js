// @flow

import type { UserId, Pie } from '../types';

import { notFound, ok } from '../helpers/http';
import { createHandler, withOptionalHttpAuthentication } from '../helpers/handlers';
import { Resource } from 'hal';

const hasNotRated = (userId: ?UserId, pie: Pie) => userId && !pie.Ratings[userId];
const isOwner = (userId: ?UserId, pie: Pie) => userId && userId === pie.UserId;
const isAbleToUploadPhoto = (userId: ?UserId, pie: Pie) => isOwner(userId, pie) && !pie.PhotoUrl;

const view = async ({ event, userId, services: { getPie } }) => {
  const { id } = event.pathParameters || {};

  const pie = await getPie(id);

  if (!pie) {
    return notFound('Unable to find the specified mince pie.');
  }

  const resource = new Resource(
    {
      id: pie.Id,
      name: pie.Name,
      rating: { avg: pie.AvgRating, total: pie.TotalRatings },
      photo: pie.PhotoUrl,
      thumbnail: pie.ThumbnailUrl,
      addedAt: new Date(pie.AddedAt).toISOString(),
    },
    `/pies/${pie.Id}`,
  );

  if (hasNotRated(userId, pie)) resource.link('rate', `/pies/${pie.Id}/rate`);
  if (isOwner(userId, pie)) resource.link('remove', `/pies/${pie.Id}`);
  if (isAbleToUploadPhoto(userId, pie)) resource.link('photo', `/pies/${pie.Id}/photo`);

  return ok(resource);
};

export default createHandler(withOptionalHttpAuthentication(view));
