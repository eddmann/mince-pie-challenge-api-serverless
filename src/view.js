/* @flow */

import * as db from './services/dynamoPieDB';
import { getUserIdFromToken } from './services/cognitoUserToken';
import view from './handlers/view';

export const handler = view({ getUserIdFromToken, db });
