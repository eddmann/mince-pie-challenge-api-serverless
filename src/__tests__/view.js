import view from '../handlers/view';

const response = () =>
  function cb(error, response) {
    if (error) fail(`Unexpected error: ${error}`);
    cb.response = response;
  };

it('views a users pie', async () => {
  const pie = {
    Id: '1',
    UserId: 'ANOTHER_USER_ID',
    Name: 'Another Mince Pie',
    AvgRating: 0,
    PhotoUrl: 'http://photo.url',
    ThumbnailUrl: 'http://thumbnail.url',
    Ratings: {},
    AddedAt: 1528217691,
  };

  const services = {
    db: { get: () => Promise.resolve(pie) },
  };

  const callback = response();

  await view(services)({ headers: {} }, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});

it('displays the photo action when no photo is present for your pie', async () => {
  const pie = {
    Id: '1',
    UserId: 'USER_ID',
    Name: 'Another Mince Pie',
    AvgRating: 0,
    PhotoUrl: undefined,
    ThumbnailUrl: undefined,
    Ratings: {},
    AddedAt: 1528217691,
  };

  const services = {
    getUserIdFromToken: () => Promise.resolve('USER_ID'),
    db: { get: () => Promise.resolve(pie) },
  };

  const callback = response();

  await view(services)({ headers: { Authorization: 'TOKEN' } }, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});

it('displays the remove action when viewing your own pie', async () => {
  const pie = {
    Id: '1',
    UserId: 'USER_ID',
    Name: 'Another Mince Pie',
    AvgRating: 0,
    PhotoUrl: 'http://photo.url',
    ThumbnailUrl: 'http://thumbnail.url',
    Ratings: {},
    AddedAt: 1528217691,
  };

  const services = {
    getUserIdFromToken: () => Promise.resolve('USER_ID'),
    db: { get: () => Promise.resolve(pie) },
  };

  const callback = response();

  await view(services)({ headers: { Authorization: 'TOKEN' } }, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});
