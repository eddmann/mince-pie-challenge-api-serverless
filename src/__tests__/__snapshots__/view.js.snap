// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the photo action when no photo is present for your pie 1`] = `
Object {
  "body": "{\\"_links\\":{\\"self\\":{\\"href\\":\\"/pies/1\\"},\\"rate\\":{\\"href\\":\\"/pies/1/rate\\"},\\"remove\\":{\\"href\\":\\"/pies/1\\"},\\"photo\\":{\\"href\\":\\"/pies/1/photo\\"}},\\"id\\":\\"1\\",\\"name\\":\\"Another Mince Pie\\",\\"rating\\":{\\"avg\\":0,\\"total\\":0},\\"addedAt\\":\\"1970-01-18T16:30:17.691Z\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/hal+json",
  },
  "statusCode": 200,
}
`;

exports[`displays the remove action when viewing your own pie 1`] = `
Object {
  "body": "{\\"_links\\":{\\"self\\":{\\"href\\":\\"/pies/1\\"},\\"rate\\":{\\"href\\":\\"/pies/1/rate\\"},\\"remove\\":{\\"href\\":\\"/pies/1\\"}},\\"id\\":\\"1\\",\\"name\\":\\"Another Mince Pie\\",\\"rating\\":{\\"avg\\":0,\\"total\\":0},\\"photo\\":\\"http://photo.url\\",\\"thumbnail\\":\\"http://thumbnail.url\\",\\"addedAt\\":\\"1970-01-18T16:30:17.691Z\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/hal+json",
  },
  "statusCode": 200,
}
`;

exports[`views a users pie 1`] = `
Object {
  "body": "{\\"_links\\":{\\"self\\":{\\"href\\":\\"/pies/1\\"}},\\"id\\":\\"1\\",\\"name\\":\\"Another Mince Pie\\",\\"rating\\":{\\"avg\\":0,\\"total\\":0},\\"photo\\":\\"http://photo.url\\",\\"thumbnail\\":\\"http://thumbnail.url\\",\\"addedAt\\":\\"1970-01-18T16:30:17.691Z\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/hal+json",
  },
  "statusCode": 200,
}
`;
